File: design/vistor/visitor.py
class Teacher(object):
    def __init__(self, name):
        self.name = name

    @property
    def data(self):
        return {"name": self.name}

    def display_json(self, visitor):
        print(visitor.to_json(self.data))


class Visitor(object):
    def __init__(self, name):
        self.name = name

    def to_json(self, data):
        import json
        data["visitor"] = self.name
        return json.dumps(data)


def test_visitor():
    t = Teacher("lao wang")
    v = Visitor("json")
    t.display_json(v)


if __name__ == "__main__":
    test_visitor()

File: design/vistor/visitor.go

package main
import (
    "fmt"
    "encoding/json"
)

type Teacher struct {
    Name string
}

func NewTeacher(name string) *Teacher {
    return & Teacher {
        Name: name,
    }
}

func (t * Teacher) Data() map[string]string {
    result := make(map[string]string)
    result["name"] = t.Name
    return result
}

func (t * Teacher) DisplayJson(v VisitorInterface) {
    data := t.Data()
    fmt.Println(v.Tojson(data))
}

type VisitorInterface interface {
    Tojson(map[string]string) string
}

type Visitor struct {
    Name string
}

func NewVisitor(name string) *Visitor {
    return & Visitor {
        Name: name,
    }
}

func (v *Visitor) Tojson(data map[string]string) string {
    data ["visitor"] = v.Name
    jsonStr, err := json.Marshal(data)
    if err != nil {
        fmt.Println("MapToJsonDemo err: ", err)
        return ""
    }
    return string(jsonStr) 
}

 
func TestVisitor() {
    t := NewTeacher("lao wang")
    v := NewVisitor("json")
    t.DisplayJson(v)
}
    
func main() {
    TestVisitor()
}
